src/cpp/r/RSexp.cpp

```diff
@@ -838,6 +854,121 @@ bool inherits(SEXP object, const char* S3Class)
    return Rf_inherits(object, S3Class);
 }
 
+std::set<std::string> makeNsePrimitives()
+{
+   std::set<std::string> nsePrimitives;
+   nsePrimitives.insert("quote");
+   nsePrimitives.insert("substitute");
+   nsePrimitives.insert("match.call");
+   nsePrimitives.insert("library");
+   nsePrimitives.insert("require");
+   nsePrimitives.insert("enquote");
+   nsePrimitives.insert("bquote");
+   nsePrimitives.insert("expression");
+   nsePrimitives.insert("evalq");
+   nsePrimitives.insert("subset");
+   return nsePrimitives;
+}
+
+const std::set<std::string>& nsePrimitives()
+{
+   static const std::set<std::string> set = makeNsePrimitives();
+   return set;
+}
+
+// Attempts to find calls to functions which perform NSE.
+bool maybePerformsNSEImpl(SEXP node,
```

Comment:
I believe this is correct. We only want to recurse if the node is a call, and we only recurse within any other sub-calls within that call. In addition, the case we care about is where we have a call and the first symbol shares a name with one of the NSE primitives (and so the 'stop state' gets wrapped in the `LANGSXP` check). The only oddity (?) is that both bits are wrapped into the same check and hence the same 'branch' of the recursion.